### ISSUE TITLE
PHP 7.3's PRCE2 error on invalid character range not reflected in explanation (previously downgraded to literal)

### ISSUE THREAD

**Issue by GreenReaper** [firasdib/Regex101#1072]
## Bug Description
PHP 7.3 uses PCRE2. This causes compilation to fail where previously an invalid character range would be ignored and treated as a literal instead. This is not reflected by the tester, which lists itself as "pcre (php)".

I appreciate that PCRE is not PCRE2; however, this distinction may not be clearly understood by all users.

### Reproduction steps
Select "pcre (php)"
Use `[^\w-_]+` as the regular expression.

### Expected Outcome
In the explanation, a notice that under certain versions of PHP there might be an issue, similar to:

> All the errors detected are listed below, from left to right, as they appear in the pattern.
> - You can not create a range with a shorthand escape sequences

[although this should probably be without the "a", or the last "s" - this is the current "python" output]

### Actual outcome

> Match a single character not present in the list below [^\w-_]+
> + Quantifier — Matches between one and unlimited times, as many times as possible, giving back as needed (greedy)
> \w matches any word character (equal to [a-zA-Z0-9_])
> -_ matches a single character in the list -_ (case insensitive)
> Global pattern flags
> i modifier: insensitive. Case insensitive match (ignores case of [a-zA-Z])

---
This is somewhat similar to #590 except that it's not an escaped character but \w means "any word character", so it simply doesn't make sense as a range.

**Comment by Ouims** [comment]
> I appreciate that PCRE is not PCRE2; however, this distinction may not be clearly understood by all users.

While this is a concern, it's definitely not a bug as PCRE2 is not supported currently.

**Comment by firasdib** [comment]
PCRE2 has made some changes that will need to be respected. The site still runs PCRE1 though, so this will have to come when I upgrade everything to PCRE2.

**Comment by Doqnach** [comment]
Would it be useful to maybe rename the current PCRE flavour to something indicating it's **not** PCRE2?
Would PCRE2 then become a new flavour as an addition, or instead of PCRE(1)?
Because it wouldn't surprise me if many tools and pieces of software will keep using PCRE(1) for a long time coming.

**Comment by firasdib** [comment]
@Doqnach They are interchangeable in 99% of the case. The purpose was to not break the API. This is however one exception.

I don't want to introduce too much confusing. There's already a lot of people who don't understand what PCRE is.

Perhaps the best would be to respect the PCRE2 API, even though we use PCRE1 for now?

**Comment by GreenReaper** [comment]
The trouble is that this regex _is_ valid in PCRE[1], just not in PRCR2. I guess it might be possible to warn of the difference in behaviour in this specific circumstance, and any others - if they are not too numerous?

This particular difference has tripped several people up judging by the PHP bug reports.

**Comment by firasdib** [comment]
@GreenReaper My suggestion is to make it invalid in PCRE1 as well, aligning it with PCRE2. It isn't intrusive, as you just have to escape a single character.

But beyond that, I don't know what the full list of differences are. Do you have a list?

**Comment by Doqnach** [comment]
The question is if you will follow PCRE2, or the PHP interpretation of it. For instance, the 'x' mode is now by default from what I read in the RFC from PHP. PHP decided to 'invert' the behaviour back to PCRE1 way to smooth out the transition. Don't know if PHP did any other changes?

**Comment by Ouims** [comment]
> It says "(php)" in the text, there's definitely something that needs adjusting here. Maybe "enhancement" comes closer than "bug", but it's definitely going to be confusing. A quickfix would be a text edit, sure.

Ahah, indeed that's a problem, however this is a non issue to me, the problem always has been that php is not a regex flavor, it should simply not say (php), as there are hundred of program that use PCRE i don't see them listed there. I don't see anything listed for js which is used in a lot of places as well.

Sure php is very common.
doesn't python also has something like that on the website, with only 2.7+ being supported or something like that? Bad idea to follow php, regex101 should not rely on what php does, but on what regex engine do. It should not say (php).

**Comment by TWiStErRob** [comment]
> regex101 should not rely on what php does, but on what regex engine do. 

I think this is going into the philosophy of the regex101 website. My quick view is that 101 suggests it's for beginners mainly with expert features. Based on this talking about engines immediately confuses beginners albeit being correct. If I made the decisions I would list most known languages/frameworks/apps and their engine in parenthesis. Supporting 4 engines sounds like it doesn't cover much, even though it supports most programming languages and many apps' search and replace.

**Comment by Ouims** [comment]
Hum... Possibly.

it is true that 101 refers to a start with something, but the website is CLEARLY not used by beginner only.
I think there are as many noob as there are intermediate/decent/expert using the website, i don't think noob represent the biggest part. I think the 101 is there mainly because it's a good name. I wasn't there when the website was created but I think "regex101" is a good name, I wouldn't be surprised if it were chosen for the name rather than for the implication that it's for noob.
And fact is there are people who are technical enough to complain about it.
I also like to think that a good website won't ever be for noob only, if anything, it's by using a good website that is made for using X (here x = regex and i mean that everyone can use the website the same way) that a noob will learn and improve, I think a lot of people improved their regex skill just by using the website and because it's made for people using regex, because it's not targeting noob only.

So, in my opinion, to be correct, it should list the engine only. Adding in parenthesis the application using the engine would only clutter the UI, but listing these apps is not a bad idea, I suggested tooltip, I think a tooltip should be used to list these apps.

**Comment by firasdib** [comment]
I don't think we have arrived at any particular conclusion here.

**Comment by Ouims** [comment]
I think there isn't a particular conclusion to reach here, points have been made.
OP is right that there is an issue, if php is not pcre1 anymore.
Aligning to pcre2 would just create confusion in a reverse fashion.

For me the solution is to remove "php" but this is your choice.

**Comment by mindplay-dk** [comment]
Is it the same issue reported here?

https://github.com/thujohn/twitter/issues/250

I ran into the same thing with an expression I wrote on regex101.com, so I took it back from code to the regex101 editor, and it was fine. I guess (based on the character offset in the error-message) at some kind of breaking change with regards to character ranges, and solved the issue by adding extra backslashes to minuses in the expression - but in PHP mode, regex101 tells me it's valid either way, so that's confusing.

> For me the solution is to remove "php" but this is your choice.

How about disambiguating the labels instead?

And adding another mode.

So for example:

  * PCRE1 (PHP 7.2 and earlier)
  * PCRE2 (PHP 7.3 and later)
  * etc.

Maybe display the engine name on the button itself, and display a tooltip on hover explaining which languages/versions make the engine available?

I don't consider myself a regex expert, but I have been using it regularly for probably 10 years.

Many of my coworkers are less experienced, so it's important to me that this remains accessible to beginners and useful to experts.

Can't we come up with a solution that works for everyone?

Simply getting a regex that doesn't work in PHP 7.3 doesn't really work for anyone.
